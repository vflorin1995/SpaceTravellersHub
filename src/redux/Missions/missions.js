import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import axios from 'axios';

const initialState = { missions: [], status: 'idle' };

export const getMissions = createAsyncThunk(
  'missions/GetMissions',
  async () => {
    try {
      const initialState = await axios.get(
        'https://api.spacexdata.com/v3/missions',
      );
      return initialState.data;
    } catch (error) {
      return error?.response;
    }
  },
);

const MissionSlice = createSlice({
  name: 'missions',
  initialState,
  reducers: {
    missionStatus: (state, { payload }) => {
      const st = state;
      st.missions = st.missions.map((obj) => (obj.mission_id === payload
        ? { ...obj, joined: !obj.joined }
        : obj));
    },
  },
  extraReducers: (Builder) => {
    Builder.addCase(getMissions.pending, (state) => ({
      ...state,
      status: 'pending',
    }))
      .addCase(getMissions.fulfilled, (state, action) => {
        const st = state;
        st.status = 'fulfilled';
        const modifiedFectedData = action.payload.map((object) => ({
          ...object,
          joined: false,
        }));
        st.missions = modifiedFectedData;
      })
      .addCase(getMissions.rejected, (state, { error }) => ({
        ...state,
        status: error,
      }));
  },
});

export const { missionStatus } = MissionSlice.actions;
export const everyMissions = (state) => state.missions;

export default MissionSlice.reducer;
