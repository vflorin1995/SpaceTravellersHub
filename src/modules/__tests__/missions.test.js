import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { HashRouter as Router } from 'react-router-dom';
import store from '../../redux/configureStore';
import Missions from '../missions';

describe('Testing by jest snapshots', () => {
  it('matches missions Snapshot', () => {
    const missionsDisplay = renderer
      .create(
        <Provider store={store}>
          <Router>
            <Missions />
          </Router>
        </Provider>,
      )
      .toJSON();
    expect(missionsDisplay).toMatchSnapshot();
  });
});
