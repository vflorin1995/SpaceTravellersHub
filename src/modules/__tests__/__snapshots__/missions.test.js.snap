// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing by jest snapshots matches missions Snapshot 1`] = `
<div
  className="w-full px-2 md:px-8 pt-4"
>
  <table
    className="w-full border"
  >
    <thead>
      <tr>
        <th
          className="font-bold capitalize text-left px-2 border py-1"
        >
          Mission
        </th>
        <th
          className="font-bold capitalize text-left px-2 border py-1"
        >
          Description
        </th>
        <th
          className="font-bold capitalize text-left px-2 border py-1"
        >
          Status
        </th>
        <th
          className="font-bold capitalize text-left px-2 border py-1"
        >
           
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;
